modules/gapi/include/opencv2/gapi/gcompiled_async.hpp

```diff
@@ -19,13 +19,17 @@ namespace cv {
 
 namespace gapi{
 namespace wip {
+    class GAsyncContext;
     //These functions asynchronously (i.e. probably on a separate thread of execution) call operator() member function of their first argument with copies of rest of arguments (except callback) passed in.
     //The difference between the function is the way to get the completion notification (via callback or a waiting on std::future object)
     //If exception is occurred during execution of apply it is transfered to the callback (via function parameter) or passed to future (and will be thrown on call to std::future::get)
-    GAPI_EXPORTS void                async(GCompiled& gcmpld, std::function<void(std::exception_ptr)>&& callback, GRunArgs &&ins, GRunArgsP &&outs);
-    GAPI_EXPORTS std::future<void>   async(GCompiled& gcmpld, GRunArgs &&ins, GRunArgsP &&outs);
-} // namespace gapi
+    GAPI_EXPORTS void                async(GAsyncContext& ctx, GCompiled& gcmpld, std::function<void(std::exception_ptr)>&& callback, GRunArgs &&ins, GRunArgsP &&outs);
```

Comment:
I recall some discussions about GAsyncContext moved to the end as optional argument.

Reply:
1. Having a default value for context parameter (and having it last), means that than there os a need to specify it, previous arguments with default parameters (mostly `GCompileArgs` for `async_apply`) have to assigned explicitly  which IMHO is inconvenient
2. Changing parameters order may break existing code, so (without extra effort) it is not an option
3. As for new - name I have two counterquestions : Why? and For example ?

Reply:
1. I find it inconvenient to support 2 interfaces (with && without context parameter) as there clearly could've been 1, but I guess there's no other obvious choice
3. It makes sense only if we decide to put "context" last: usually a context is something that other stuff depends on, so it is passed as a first parameter and called, well, a context. If it's optional then it's not really a context (which seems to always exist by definition), it's rather an additional (meta?-)information that may or may not be present.

Anyway, I'm totally fine with naming it `GAsyncContext` as long as we have it as a first parameter, just thinking that if it's last && optional, context is not really the name it should have

Reply:
+1 to make it last. It is a common vision now (GVA asks for the same)

Reply:
From our past notes:

>    - Default contexts - will be one per operation

Please implement the agreed solution.

Reply:
In fact, It is implemented - then no context is specified - operation uses devoted "dummy" context which is not exposed in any way. So, there is no way to cancel all operations that did not specified contexts explicitly.